feat: additional `atmos docs` parameters for specifying width, using auto-styling and color profile, and preserving new lines

RoseSecurity opened the pull request:
## what

![atmos_docs](https://github.com/user-attachments/assets/24b098eb-d7e1-4d19-bc84-5909ac34a672)

- Add an additional `atmos docs` flag for specifying the width of markdown output
- Utilizing auto-styling based on light or dark mode preferences instead of hardcoding to dark
- Preserving new lines with rendered markdown

## why

- Enhance the user experience for interacting with documentation. The `width` flag is useful for users who prefer seeing wider output for Terraform docs-generated tables

## references

- [`glow` docs](https://github.com/charmbracelet/glow/tree/master)

RoseSecurity commented on cmd/docs.go:
```diff
@@ -95,4 +104,5 @@ var docsCmd = &cobra.Command{
 
 func init() {
 	RootCmd.AddCommand(docsCmd)
+	docsCmd.Flags().UintVarP(&width, "width", "w", 80, "word-wrap at width")
```
How would you define reasonable bounds for wider output across varying terminal sizes and user preferences?

osterman replied:
I think the key is to make it responsive, like in `glow` and the recommendation above.

https://github.com/charmbracelet/glow/blob/master/main.go#L170

RoseSecurity replied:
Just played around with this recommendation and dynamically setting the width based on terminal settings makes for a really nice experience

RoseSecurity commented on cmd/docs.go:
```diff
@@ -95,4 +123,5 @@ var docsCmd = &cobra.Command{
 
 func init() {
 	RootCmd.AddCommand(docsCmd)
+	docsCmd.Flags().UintVarP(&width, "width", "w", 0, "word-wrap at width (set to 0 to disable)")
```
Good call!

RoseSecurity commented on cmd/docs.go:
```diff
@@ -37,18 +39,33 @@ var docsCmd = &cobra.Command{
 				u.LogErrorAndExit(schema.CliConfiguration{}, err)
 			}
 
+			// Detect terminal width if width flag is not set
+			width, _ := cmd.Flags().GetUint("width")
+			if !cmd.Flags().Changed("width") {
+				if term.IsTerminal(int(os.Stdout.Fd())) {
+					w, _, err := term.GetSize(int(os.Stdout.Fd()))
+					if err == nil {
+						width = uint(w)
+					}
+
+					if width > 120 {
+						width = 120
+					}
+
+					if width == 0 {
+						width = 80
+					}
+				}
+			}
```
I am content with the compact readability and maintainability of the code

osterman commented on cmd/docs.go:
```diff
@@ -95,4 +122,5 @@ var docsCmd = &cobra.Command{
 
 func init() {
 	RootCmd.AddCommand(docsCmd)
+	docsCmd.Flags().UintP("width", "w", 0, "Set word-wrap width (0 disables word wrapping). Influences output formatting, particularly for tables.")
```
Seems more appropriate for `docs.width` setting in `atmos.yaml`

RoseSecurity replied:
Would you envision it under the `settings` block or a separate `docs` structure?

```yaml
docs:
  width: 140
  # And eventually...
  pagination: (false | true)
```

osterman replied:
Should it be max-width? Since width is otherwise dynamic?

osterman replied:
Or if width is "auto" or undefined it's responsive, and otherwise constant.

RoseSecurity replied:
Just added the blocks to the `atmos.yaml`!
```yaml
settings:
  docs:
    max-width: 140
```

This is my first stab at this kind of change so just let me know of any other areas you'd like updated! I tested it with our configuration and it works as intended

osterman commented on cmd/docs.go:
```diff
@@ -65,7 +82,17 @@ var docsCmd = &cobra.Command{
 			}
 
-			componentDocs, err := glamour.Render(string(readmeContent), "dark")
+			r, err := glamour.NewTermRenderer(
+				glamour.WithColorProfile(lipgloss.ColorProfile()),
+				glamour.WithAutoStyle(),
+				glamour.WithPreservedNewLines(),
+				glamour.WithWordWrap(int(width)),
+			)
+			if err != nil {
+				u.LogErrorAndExit(schema.CliConfiguration{}, err)
+			}
+
```
This seems like a more descriptive error message

osterman commented on cmd/docs.go:
```diff
@@ -37,18 +39,33 @@ var docsCmd = &cobra.Command{
+						width = int(w)
+					}
+
+					if width > 120 {
+						width = 120
+					}
```
Seems like we're not using max width?
```suggestion
					if w > width {
						w = width
					}
```

RoseSecurity replied:
Updated the naming!

osterman commented on cmd/docs.go:
```diff
@@ -37,18 +39,33 @@ var docsCmd = &cobra.Command{
 			}
 
+			// Detect terminal width if width flag is not set
+			width := cliConfig.Settings.Docs.MaxWidth
```
We should probably rename this to maxWidth to better reflect it's current usage.
```suggestion
			width := cliConfig.Settings.Docs.MaxWidth
```

RoseSecurity replied:
Updated the naming!

osterman commented on cmd/docs.go:
```diff
@@ -37,18 +39,33 @@ var docsCmd = &cobra.Command{
+						maxWidth = int(w)
+					}
+
+					if maxWidth > 120 {
+						maxWidth = 120
+					}
```
The 120 max width seems arbitrary and we're not capping the width.

```suggestion
					if w > maxWidth {
						w = maxWidth
					}
```

RoseSecurity replied:
120 characters is a common standard width limit in terminal applications. It's wide enough to display most content comfortably while avoiding excessively long lines that become hard to read. This is how `glow` works as well

osterman replied:
I don't disagree it's a reasonable default, if no default is specified. As implemented, however, `maxWidth` is not the maxWidth, and there's an implicit max capped on the maxWidth, that seems awkward to me.

I think a conventional way of handling this would be:

1. If `maxWidth` is specified and greater than zero, we should set the width to the smaller (`min`) of screenWidth and maxWidth. Otherwise, if maxWidth is unset or zero, we should default to the `screenWidth`, capped at 120 as you have.
2. Remove max maxWidth to avoid adding an opinionated upper constraint over maxWidth, as this limits the flexibility of maxWidth to dynamically adjust based on the context.

osterman commented on cmd/docs.go:
```diff
@@ -37,18 +39,33 @@ var docsCmd = &cobra.Command{
+						maxWidth = 120
+					}
+
+					if maxWidth == 0 {
+						maxWidth = 80
+					}
```
This will need to move higher, if this other suggestion is accepted

https://github.com/cloudposse/atmos/pull/757/files#r1828219437

osterman commented on cmd/docs.go:
```diff
@@ -37,18 +39,33 @@ var docsCmd = &cobra.Command{
+			maxWidth := cliConfig.Settings.Docs.MaxWidth
+			if maxWidth == 0 {
+				if term.IsTerminal(int(os.Stdout.Fd())) {
+					w, _, err := term.GetSize(int(os.Stdout.Fd()))
```
In this context `w` is overly vague. I think it would be more clear if we called it screenWidth and separetly define the desired width as the `min(screenWidth, maxWidth, 120)`

```suggestion
					screenWidth, _, err := term.GetSize(int(os.Stdout.Fd()))
```

RoseSecurity replied:
Just refactored the logic to keep `screenWidth` at 120 unless the conditions specify otherwise, making it more predictable and flexible

osterman commented:
Unfriendly error message: (this is more debug log level)

Component folder does not exist. We should instead say, "Component vpc not found"
<img width="775" alt="image" src="https://github.com/user-attachments/assets/a52608b9-9f81-4971-abc8-b73445594d2d">

Good error message:

(no `README.md` file exists in a valid component folder)

<img width="509" alt="image" src="https://github.com/user-attachments/assets/c00cdaa0-b397-4298-8047-109cc48dfe86">

osterman commented:
It does not appear to be detecting the full width of my terminal. Maybe we should add some debug messages to see what the max width is determined as and what width is selected.

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/4ce68f96-fc66-4e17-be66-e1f31a3a1cd7">

Note, the screen is much wider, and the text is not wrapping and is truncated.

RoseSecurity commented:
> It does not appear to be detecting the full width of my terminal. Maybe we should add some debug messages to see what the max width is determined as and what width is selected.
> 
> <img alt="image" width="1512" src="https://private-user-images.githubusercontent.com/52489/383242605-4ce68f96-fc66-4e17-be66-e1f31a3a1cd7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5OTgwODEsIm5iZiI6MTczMDk5Nzc4MSwicGF0aCI6Ii81MjQ4OS8zODMyNDI2MDUtNGNlNjhmOTYtZmM2Ni00ZTE3LWJlNjYtZTFmMzFhM2ExY2Q3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA3VDE2NDMwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxYmI1NGQ1NjZhZjMzOTRhNmUxYzUwOTBjOGI4ZDRjYmY4YzA3YTJiZmI4MjYwZjJlNTI1Y2U1YzFhMjQ0Y2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wZtxl7aqPW8ekspPFuNX8db08YRPXWF60q0okQLIrS8">
> Note, the screen is much wider, and the text is not wrapping and is truncated.

Just removed the logic that defaults back to `120` for the maximum width, allowing the word wrap to dynamically adjust to the terminal size. I think it looks a little less clean when rendered this way but feel free to check it out

osterman commented:
It appears there's some fudge factor needed with the width calculation. It's off by 2 characters. I'm wondering if it's related to something like `\r\n`

![CleanShot 2024-11-07 at 11 39 51](https://github.com/user-attachments/assets/ea08e520-8de6-4a6b-8660-eb6fea87c3e5)

>  I think it looks a little less clean when rendered this way but feel free to check it out

Maybe this is what you're referring to. But this is a bug, not a result of limiting the width. That causes an extra line feed between each row.

If I expand my screen by 2 characters, I get
<img width="1502" alt="image" src="https://github.com/user-attachments/assets/99d7ff1a-4707-4be5-bd8a-4ce2ca73ce08">


Which is clean.

RoseSecurity commented:
> It appears there's some fudge factor needed with the width calculation. It's off by 2 characters. I'm wondering if it's related to something like `\r\n`
> 
> ![CleanShot 2024-11-07 at 11 39 51](https://private-user-images.githubusercontent.com/52489/384077841-ea08e520-8de6-4a6b-8660-eb6fea87c3e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzM2MDMsIm5iZiI6MTczMTMzMzMwMywicGF0aCI6Ii81MjQ4OS8zODQwNzc4NDEtZWEwOGU1MjAtOGRlNi00YTZiLTg2NjAtZWI2ZmVhODdjM2U1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEzNTUwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2Y2IwMmM2NjljNTU0MDMxOTA4YzBhOTdjNjc5ZWMyZDI1MjgzNWU3MjY4NjgxNTRmOTY2M2EwOWRlZDcyMjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xsmW5xHtzjZS6psihrFqoywNuHk3OkuhAUNEJcWDWBM)
> 
> > I think it looks a little less clean when rendered this way but feel free to check it out
> 
> Maybe this is what you're referring to. But this is a bug, not a result of limiting the width. That causes an extra line feed between each row.
> 
> If I expand my screen by 2 characters, I get <img alt="image" width="1502" src="https://private-user-images.githubusercontent.com/52489/384078329-99d7ff1a-4707-4be5-bd8a-4ce2ca73ce08.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzM2MDMsIm5iZiI6MTczMTMzMzMwMywicGF0aCI6Ii81MjQ4OS8zODQwNzgzMjktOTlkN2ZmMWEtNDcwNy00YmU1LWJkOGEtNGNlMmNhNzNjZTA4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEzNTUwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5MzIyMjU5N2IzNDJlZDc2NDI2MWU1NWU0MDBjNjcwYmQ5NTE5NThlY2I0ODg5YTM2ZGFiNTI4MDMwYjM2YjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gDzz2-u7j4HtmQnKQ4sxBfpGdnyBfgw_goia2RTX2gM">
> 
> Which is clean.

Thanks for the great find! Interesting that there is a subtle padding effect at the terminal boundaries like that. I adjusted accordingly and added a comment for the width subtraction in the code!

osterman commented:
Perfect, LGTM! 

<img width="1551" alt="image" src="https://github.com/user-attachments/assets/8a37f55e-5acb-40da-8c5f-48df7131d576">

osterman commented:
Regarding the incorrect screen width calculations (which we resolved with a hack), it might be related to this recently opened issue.

- charmbracelet/bubbletea#1225

M0hammadUsman commented:
> Regarding the incorrect screen width calculations (which we resolved with a hack), it might be related to this recently opened issue.
> 
> - charmbracelet/bubbletea#1225

@osterman It has been resolved in bubbletea 1.2.1, now the issue is still open there because there is something else. 

See my last comment on that issue to get more insight.
